dj/api/nodes.py

```diff
@@ -113,17 +146,208 @@ def validate_node(
+        data.name,
+        node_revision.version,
+        node_revision.parents,
+    )
+
+    # Infer the new node revision's column names and column types by inspecting the parent nodes
+    try:
+        node_revision.columns = infer_columns(
+            node_revision.query,
+            parents=node_revision.parents,
+        )
+    except Exception as exception:  # pylint: disable=broad-except
+        error_message = f"Cannot infer columns for {data.name}: {str(exception)}"
+        _logger.error(error_message)
+    return node_revision
+
+
+@router.post("/nodes/", response_model=RefNodeMetadata)
+def create_node(
+    data: Union[CreateSourceNode, CreateNode],
+    session: Session = Depends(get_session),
+) -> RefNodeMetadata:
+    """
+    Create a node.
+    """
+    query = select(ReferenceNode).where(ReferenceNode.name == data.name)
+    ref_node = session.exec(query).one_or_none()
+    if ref_node:
+        raise DJException(f"A node with name `{data.name}` already exists.")
+
+    ref_node = ReferenceNode(name=data.name, type=data.type, current_version=0)
+
+    if data.type == NodeType.SOURCE:
+        node_revision = create_source_node_revision(data, session)
+    else:
+        node_revision = create_node_revision(data, session)
+
+    # Point the reference node to the new node revision.
+    node_revision.reference_node = ref_node
+    node_revision.version = ref_node.current_version + 1
+    ref_node.current_version = node_revision.version
+
+    node_revision.extra_validation()
+
+    session.add(ref_node)
+    session.commit()
+    session.refresh(ref_node)
+    return ref_node  # type: ignore
+
+
+@router.put("/nodes/{name}/", response_model=RefNodeMetadata)
+def update_node(
+    name: str,
+    data: Union[CreateSourceNode, NodeBase],
+    *,
+    session: Session = Depends(get_session),
+) -> RefNodeMetadata:
+    """
+    Update a node.
+    """
+    query = (
+        select(ReferenceNode)
+        .where(ReferenceNode.name == name)
+        .with_for_update()
+        .execution_options(populate_existing=True)
+    )
+    ref_node = session.exec(query).one_or_none()
+    if not ref_node:
+        raise DJException(
+            message=f"A node with name `{name}` does not exist.",
+            http_status_code=404,
+        )
+
+    # Verify that the changes are valid
+    if ref_node.name != data.name:
+        raise DJException(message="Cannot change node name!")
+    if ref_node.type != data.type:
+        raise DJException(message="Cannot change node type!")
+
+    # If nothing has changed, do not create the new node revision
+    if (
+        ref_node.current.query == data.query
+        and ref_node.current.description == data.description
+        and ref_node.current.mode == data.mode
+    ):
+        return ref_node
+
+    # Build the new version and find the new node version's parents
+    if data.type == NodeType.SOURCE:
+        node_revision = create_source_node_revision(data, session)
```

Comment:
Adding the `UpdateNode` class without the name means you'll have to inject the name here before calling this function.
```py
node_revision = create_source_node_revision(CreateNode(name=ref_node.name, **data.dict()), session)
```